ledger/vm/vm.go

```diff
@@ -68,6 +71,177 @@ func TransferTheta(db StateDB, sender, recipient common.Address, amount *big.Int
 	db.AddThetaBalance(recipient, amount)
 }
 
+func parseBLSSummary(summary []byte) (holderAddress common.Address, blsPubkey *bls.PublicKey, blsPop *bls.Signature, holderSig *crypto.Signature, ok bool) {
+	if len(summary) != 229 && len(summary) != 261 {
+		return
+	}
+
+	holderAddress = common.BytesToAddress(summary[:20])
+	blsPubkey, err := bls.PublicKeyFromBytes(summary[20:68])
+	if err != nil {
+		return
+	}
+	blsPop, err = bls.SignatureFromBytes(summary[68:164])
+	if err != nil {
+		return
+	}
+	holderSig, err = crypto.SignatureFromBytes(summary[164:])
+	if err != nil {
+		return
+	}
+
+	ok = true
+	return
+}
+
+func CheckBLSSummary(summary []byte) bool {
+	guardianAddr, blsPubkey, blsPop, holderSig, ok := parseBLSSummary(summary)
+	if !ok {
+		return false
+	}
+
+	return checkBlsSummary(blsPubkey, blsPop, holderSig, guardianAddr)
+}
+
+func checkBlsSummary(blsPubkey *bls.PublicKey, blsPop *bls.Signature, holderSig *crypto.Signature, guardianAddr common.Address) bool {
+	if blsPubkey.IsEmpty() {
+		return false
+	}
+	if blsPop.IsEmpty() {
+		return false
+	}
+	if holderSig == nil || holderSig.IsEmpty() {
+		return false
+	}
+
+	if !holderSig.Verify(blsPop.ToBytes(), guardianAddr) {
+		return false
+	}
+
+	if !blsPop.PopVerify(blsPubkey) {
+		return false
+	}
+
+	return true
+}
+
+// StakeToGuardian stake Theta to given guardian node.
+func StakeToGuardian(db StateDB, sender common.Address, guardianSummary []byte, amount *big.Int) bool {
+	if amount.Cmp(core.MinGuardianStakeDeposit) < 0 {
+		return false
+	}
+	if db.GetThetaBalance(sender).Cmp(amount) < 0 {
+		return false
+	}
+
+	guardianAddr, blsPubkey, blsPop, holderSig, ok := parseBLSSummary(guardianSummary)
+	if !ok {
+		return false
+	}
+
+	view := db.(*state.StoreView)
+	gcp := view.GetGuardianCandidatePool()
+	if !gcp.Contains(guardianAddr) {
+		if !checkBlsSummary(blsPubkey, blsPop, holderSig, guardianAddr) {
+			return false
+		}
+	}
+
+	db.SubThetaBalance(sender, amount)
+
+	err := gcp.DepositStake(sender, guardianAddr, amount, blsPubkey, view.GetBlockHeight())
+	if err != nil {
+		return false
+	}
+
+	view.UpdateGuardianCandidatePool(gcp)
+
+	return true
+}
+
+// UnstakeFromGuardian unstake from Guardians.
+func UnstakeFromGuardian(db StateDB, addr common.Address, guardianAddr common.Address) bool {
+	view := db.(*state.StoreView)
+	gcp := view.GetGuardianCandidatePool()
+	currentHeight := view.GetBlockHeight()
```

Comment:
tx_withdraw_stake.go was using exec.state.Height() as the current height (which is actually off-by-one). Do we want to keep the behavior consistent (although view.GetBlockHeight() returns the correct height)?

https://github.com/thetatoken/theta-protocol-ledger/blob/583dea58dffe61691a593238145038a63bdc738e/ledger/execution/tx_withdraw_stake.go#L100

tx_deposit_stake.go was using the correct heights though.

Reply:
updated